spring-cloud-gcp-bigquery/src/test/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandlerIntegrationTests.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.bigquery.integration;
+
+import java.io.File;
+import java.util.HashMap;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.bigquery.FormatOptions;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobInfo.WriteDisposition;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableResult;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.MessageBuilder;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assume.assumeThat;
+
+public class BigQueryFileMessageHandlerIntegrationTests {
+
+	private static final String DATASET_NAME = "test_dataset";
+
+	private static final String TABLE_NAME = "test_table";
+
+	private BigQuery bigquery;
+
+	private BigQueryFileMessageHandler messageHandler;
+
+	@BeforeClass
+	public static void prepare() {
+		assumeThat(
+				"BigQuery integration tests are disabled. "
+						+ "Please use '-Dit.bigquery=true' to enable them.",
+				System.getProperty("it.bigquery"), is("true"));
+	}
+
+	@Before
+	public void setup() {
+		this.bigquery = BigQueryOptions.getDefaultInstance().getService();
+
+		this.messageHandler = new BigQueryFileMessageHandler(this.bigquery);
+		this.messageHandler.setWriteDisposition(WriteDisposition.WRITE_EMPTY);
+
+		// Clear the previous dataset before beginning the test.
+		this.bigquery.delete(TableId.of(DATASET_NAME, TABLE_NAME));
+	}
+
+	@Test
+	public void testMessageHandlerLoadFile() throws InterruptedException {
+		HashMap<String, Object> messageHeaders = new HashMap<>();
+		messageHeaders.put(BigQuerySpringMessageHeaders.DATASET_NAME, "test_dataset");
+		messageHeaders.put(BigQuerySpringMessageHeaders.TABLE_NAME, "test_table");
+		messageHeaders.put(BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.csv());
+
+		Message<File> message = MessageBuilder.createMessage(
+				new File("src/test/resources/data.csv"),
+				new MessageHeaders(messageHeaders));
+
+		Job job = (Job) this.messageHandler.handleRequestMessage(message);
+		job.waitFor();
```

Comment:
I think we need to have some logic in the `BigQueryFileMessageHandler` to handle this `Job` internally. Something like `sync` and `sendTimeout` options and some async handling otherwise.
See https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java for more info.

Reply:
OK. I have a question though -

This `Job` object is something returned by the Google Client libs.

If you want to wait for the job to complete **sync**, you call:
```
job.waitFor()

// Then you can start using the object
job.getResults().getRowsUpdated()
```

If you want to do this **async**, you have to do some kind of polling like this:
```
while (!JobStatus.State.DONE.equals(job.getStatus().getState())) {
  Thread.sleep(1000L);
  job = job.reload();
}

// Then you can start using the object
job.getResults().getRowsUpdated()
```

What is the recommended practice in Spring Integration for abstracting polling logic from the user?

Reply:
Well, it still sounds like a background task and a `ListenableFuture` as a return from this `handleRequestMessage`.
A `SettableListenableFuture` should be used internally for populating the value.

Reply:
Gotcha sounds good. In that case, would I also need to add a thread pool responsible for polling these jobs? Or would the user provide this?

Reply:
We can have a default one, but still allow end-user to inject a managed one. We don't have such a sample though to share because we have always relied on the async execution in the target client, but essentially, it doesn't matter for end-user who is responsible for task execution.